game-core/src/main/java/games/strategy/engine/data/GameParser.java

```diff
@@ -1204,72 +1256,124 @@ private void parseRepairFrontierRules(final List<Element> elements, final Repair
+    final List<Tuple<String, String>> results = new ArrayList<>();
+    for (final Element option : options) {
       // decapitalize the property name for backwards compatibility
-      final String name = decapitalize(current.getAttribute("name"));
+      final String name = decapitalize(option.getAttribute("name"));
       if (name.isEmpty()) {
-        throw newGameParseException("Option name with zero length");
+        throw newGameParseException("Option name with zero length for attachment: " + attachment.getName());
+      }
+      final String value = option.getAttribute("value");
+      if (containsEmptyForeachVariable(value, foreach)) {
+        continue; // Skip adding option if contains empty foreach variable
+      }
+      final String valueWithForeach = replaceForeachVariables(value, foreach);
+      final String count = option.getAttribute("count");
+      String optionValues = (!count.isEmpty() ? count + ":" : "") + valueWithForeach;
+      if (!variables.isEmpty()) {
+        final List<String> listWithVariables = new ArrayList<>();
+        for (final String s : Splitter.on(':').split(optionValues)) {
+          listWithVariables.add(variables.containsKey(s) ? Joiner.on(":").join(variables.get(s)) : s);
+        }
+        optionValues = Joiner.on(":").join(listWithVariables);
       }
-
-      final String value = current.getAttribute("value");
-      final String count = current.getAttribute("count");
-      final String itemValues = (count.length() > 0 ? count + ":" : "") + value;
       try {
         attachment.getProperty(name)
             .orElseThrow(() -> newGameParseException(String.format(
                 "Missing property definition for option '%s' in attachment '%s'",
                 name, attachment.getName())))
-            .setValue(itemValues);
+            .setValue(optionValues);
       } catch (final GameParseException e) {
         throw e;
       } catch (final Exception e) {
-        throw newGameParseException("Unexpected Exception while setting values for attachment" + attachment, e);
+        throw newGameParseException("Unexpected Exception while setting values for attachment: " + attachment, e);
       }
 
-      options.add(Tuple.of(name, itemValues));
+      results.add(Tuple.of(name, optionValues));
+    }
+    return results;
+  }
+
+  private String replaceForeachVariables(final String s, final Map<String, String> foreach) {
+    String result = s;
+    for (final Entry<String, String> entry : foreach.entrySet()) {
+      result = result.replace(entry.getKey(), entry.getValue());
+    }
+    return result;
+  }
+
+  private boolean containsEmptyForeachVariable(final String s, final Map<String, String> foreach) {
+    for (final Entry<String, String> entry : foreach.entrySet()) {
+      if (entry.getValue().isEmpty() && s.contains(entry.getKey())) {
+        return true;
+      }
     }
-    return options;
+    return false;
```

Comment:
So if I understand this correctly the only reason this code iterates over the map is because we can't be sure if we expect `VariableList` or `$VariableItem$` at this point, right?
(Besides the fact I dislike the fact that there isn't any variable identifier at all, but whatever, not a map maker:) Wouldn't it be better to just do
```java
return foreach.getOrDefault(s, " ").isEmpty() || foreach.getOrDefault("$" + s + "$", " ").isEmpty();
```
or even
```java
return foreach.getOrDefault(s, foreach.getOrDefault("$" + s + "$", " ")).isEmpty();
```
or even check what to expect beforehand and pass an s with the appropriate format?

Reply:
@ron-murhammer Bump

Reply:
Ok. I need to make some adjustments based on moving forward with $ and @ as variable identifiers so will take a look at this.

Reply:
No, this is actually checking if the current option contains the 'variableElement' not equals it (you could probably use a regex). So you could have something like this:
```
<option name="requiresUnit" value="@Player@Barracks"/>
```
```
<variable name="Player">
  <element>German</element>
  <element></element>
</variable>
```

So it generates this for the first iteration: 
```
<option name="requiresUnit" value="GermanBarracks"/>
```

And skips adding the option at all for the second iteration.

Reply:
Thanks for the clarification (although I'm not convinced your example will produce a valid string)